custom_components/tuya_local/devices/cecotec_readywarm_8400.yaml

```diff
@@ -15,15 +15,14 @@ primary_entity:
         - dps_val: true
           constraint: preset_mode
           conditions:
-            - dps_val: "null"
-              value: fan_only
-              hidden: true
```

Comment:
This  is necessary for the Fan only mode to work.  With this change in place - i cant engage the fan only mode.

Reply:
I get this error in the log:
2024-12-28 10:12:52.186 WARNING (MainThread) [custom_components.tuya_local.climate] cecotec_readywarm.yaml/climate: Unrecognised HVAC Mode of True ignored

and on screen i get this error:
![Screenshot 2024-12-28 at 10 12 52](https://github.com/user-attachments/assets/d1a6e7f6-b7e5-4b25-a2cf-ec0a5bf1225c)

Reply:
SInce it is marked as hidden, it cannot be related to engaging fan only mode, since the mapping for "off" is used in that case, which is still there.

Reply:
> Since it is marked as hidden, it cannot be related to engaging fan only mode, since the mapping for "off" is used in that case, which is still there.

Ok, but thats the behaviour I am seeing.  If I put the Heater into fan only mode withe remote control manually the screen will eventually update correctly - but using the home assistant UI leads to the above behaviour.

Reply:
I have just noticed that eventually the UI will correct itself eventually, I am assuming when the state is pushed from the Device back to home assistant.  But it goes into unknown state with errors in the log until that happens.

Reply:
Also just another thing - I had struggled with getting this right until I saw another example of a fan heater working this way: vtaplus_axial_fanheater.yaml

Reply:
Can you capture the debug log of changing from heat to fan_only mode?
I think if the sequence of what the device is actually doing can be understood, it will confirm whether this is a correct workaround or a placebo.

Reply:
Yep, no problem. Apologies for the delay.  See below the log when I change the heater to fan only.



`
`2024-12-30 10:51:14.946 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-30 10:51:14.946 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Considering mapping for range of preset_mode
2024-12-30 10:51:14.946 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 to True from fan_only
2024-12-30 10:51:14.946 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Considering mapping for range of hvac_mode
2024-12-30 10:51:14.946 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater new pending updates: {"4": {"value": "none", "updated_at": 1735555874.9468095, "sent": false}, "1": {"value": true, "updated_at": 1735555874.9468095, "sent": false}}
2024-12-30 10:51:14.950 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater sending dps update: {"1": true, "4": "none"}
2024-12-30 10:51:15.077 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater received {"1": true, "full_poll": true}
2024-12-30 10:51:15.079 WARNING (MainThread) [custom_components.tuya_local.climate] cecotec_readywarm.yaml/climate: Unrecognised HVAC Mode of True ignored
2024-12-30 10:51:15.080 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-30 10:51:15.080 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-30 10:51:15.082 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-30 10:51:15.082 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-30 10:51:15.190 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater received {"1": true, "2": 25, "3": 22, "4": "off", "8": false, "19": "0h", "20": 0, "full_poll": false}
2024-12-30 10:51:15.192 WARNING (MainThread) [custom_components.tuya_local.climate] cecotec_readywarm.yaml/climate: Unrecognised HVAC Mode of True ignored
2024-12-30 10:51:15.193 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-30 10:51:15.193 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-30 10:51:15.193 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-30 10:51:15.194 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel`

Reply:
'Fan Heater sending dps update: {"1": true, "4": "none"}'

If I am reading this log message correctly then this line seems to be wrong.  
The DP 4 value being sent to the fan should be "off" not "none".

Reply:
If I take your updates as you committed them but change line 44 to unhide the off/none preset things work fine.  And  I can see the correct dp value of "off" being sent to the heater.  See log below.

I did see some logic in the climate entity definition around different behaviour if the dp was hidden or not.  Maybe there is something there?

`2024-12-30 11:43:05.030 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-30 11:43:05.031 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 to off from none
2024-12-30 11:43:05.031 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Considering mapping for range of preset_mode
2024-12-30 11:43:05.031 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 to True from fan_only
2024-12-30 11:43:05.031 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Considering mapping for range of hvac_mode
2024-12-30 11:43:05.031 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater new pending updates: {"4": {"value": "off", "updated_at": 1735558985.031372, "sent": false}, "1": {"value": true, "updated_at": 1735558985.031372, "sent": false}}
2024-12-30 11:43:05.033 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater sending dps update: {"1": true, "4": "off"}
2024-12-30 11:43:05.151 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater received {"1": true, "full_poll": true}
2024-12-30 11:43:05.152 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-30 11:43:05.153 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-30 11:43:05.154 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 value from True to fan_only
2024-12-30 11:43:05.154 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 value from True to fan_only
2024-12-30 11:43:05.154 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-30 11:43:05.155 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-30 11:43:05.155 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-30 11:43:05.155 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-30 11:43:05.264 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater received {"1": true, "2": 25, "3": 20, "4": "off", "8": false, "19": "0h", "20": 0, "full_poll": false}
2024-12-30 11:43:05.265 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-30 11:43:05.266 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-30 11:43:05.267 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 value from True to fan_only
2024-12-30 11:43:05.267 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 value from True to fan_only
2024-12-30 11:43:05.267 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-30 11:43:05.267 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-30 11:43:05.267 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-30 11:43:05.268 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off`

Reply:
The code dealing with hidden values does not seem to be working correctly, however I cannot see how it is ending up with this result.

What seems to be happening is that the hidden mapping seems to be found first, but in that case it should later get filtered out, and the original value (fan_only) should be sent to the dp - but since this is a conditional mapping, it may be more complex than that.

The solution should be to remove that mapping entirely.  If the behaviour is not correct with that, then the debug logs from that case should point to what is going on.  It seems that the device returns a "full_poll" result with only the power indicator, then the full data comes through later, but that might just be a side effect of the wrong value being sent to the device immediately before that.

Reply:
I removed the Off Preset as per your suggestion and it appears to work fine.  Thanks for that.